packages/compiler/src/core/checker.ts

```diff
@@ -2778,6 +2790,164 @@ export function createChecker(program: Program): Checker {
     return (resolved?.declarations.filter((n) => isTemplatedNode(n)) ?? []) as TemplateableNode[];
   }
 
+  function getReferencedModel(
+    propertyNode: ObjectLiteralPropertyNode | ModelPropertyNode
+  ): Model | undefined {
+    const ARRAY_MODEL_FLAG = "@@array_model_flag@@";
+
+    if (propertyNode.kind === SyntaxKind.ModelProperty) {
+      return getReferencedModelFromDecoratorArgument(propertyNode);
+    } else if (propertyNode.kind === SyntaxKind.ObjectLiteralProperty) {
+      let r = getReferencedModelFromScalarConstructor(propertyNode);
+      r ??= getReferencedModelFromConstAssignment(propertyNode);
+      r ??= getReferencedModelFromDecoratorArgument(propertyNode);
```

Comment:
feels like this could be traversed at the same time no?

Reply:
Feels all those funcitons are nearly the same execpt the condition to find the parent, why not just 
```
findAncestor(propertyNode, node => node.kind === ...)
```

Reply:
all these method is similar, step1 get ancestor, step2 figure out the type, step3, return the type according to the path. but the step1 and 2 are different and only step3 is the same which is only one line. please let me know if there is good way to make the step2 merged.

I put them in 3 methods to make logic clean and focusing. it's possible to traverse them together but will be many if-else and be harder to read and understand. Also it's possible to have an overall step1-step2-step3 main method and passing detail step implementation as argument, but it feels a little overkill for me here.